accepted/0000-yomi-initial-integration.md

```diff
@@ -0,0 +1,363 @@
+
+This will imply that the Yomi state is outside the highstate, and that
+this new API needs to be designed.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Some of the open questions that I am able to identify, are:
+
+## How the minion can find the master automatically.
+
+The current Yomi image (PXE boot and ISO), provides a systemd
+overwrite that is executed before the salt-minion service. This piece
+of code will set a minion ID for the system based on some rules:
+
+- Use the ID injected via the bootloader parameter 'minion='
+
+- If a FQDN is provided via DHCP, this name will be used as the ID
+
+- If no parameter or of the hostname is localhost, the MAC address
+  will be used as minion ID
+
+The question is to validate those rules for the general use case.
+
+## How to find the master node.
+
+In a similar way, an systemd overlay will take care of configuring
+where to find the salt-master node.
+
+- If the bootloader parameter 'master=' is provided, a new file
+  'master.conf' will be added inside '/etc/salt/minion.d'
+
+- If not, the default 'master' name will be used
+
+Is there a way to use DHCP to inject the correct address for the
+salt-master node?
+
+## Disable SSH by default.
+
+The current PXE Boot and ISO image contains a root user than can be
+accessed without password. The problem is that as the SSH service is
+enabled, you can remote login into the new system as root without
+password.
+
+One solution is to disable password-less login via SSH, and inject via
+DHCP some parameters that can enable the login, or inject authorized
+certificates.
+
+## Simplified forms
+
+The prototype provide a set of subforms that cover all the pillar data
+that Yomi can understand. Maybe there is the option of hidding those
+forms and provide a simplified version, based of a single option like:
+(1) what kind of OS want to be installed, (2) simplified partition
+definition.
+
+The list of packages, repositories, service configuration and other
+elements will be described and configured by current parts of Uyuni,
+after the first boot.
+
+## Workflow
+
+In the 'Detailed design' section we pointed a problem with the naive
+workflow, and two alternatives were provided. Which one to choose, and
+how to implement it, is something that needs to be clarified before
+the approval of this RFC.
+
+## Overlap with different pillar and formula
+
+The pillar data for Yomi requires the definition of the language,
+keyboard layout to use, timezone, list of packages, etc. This can
+overlap with other formulas like 'locate', or some other components of
+Uyuni like 'Sofware'
+
+How to resolve this overlap is still an open question.
+
+## Be ready for Redfish integration
+
+In the future maybe we will want to integrate more different ways of
+installation, like Redfish, where Yomi can have a role on
+it. Decisions done in this stage can influence the possibilities that
+we will have in the future.
```

Comment:
In the first iteration, I don't expect too many complications with Redfish. Unless I'm missing something, Redfish will mainly allow for three things that could be useful for us in the Yomi scenario:

1. http(s) boot (technically not a Redfish feature, but all servers that support Redfish also have an http(s) boot enabled UEFI firmware, and you can use Redfish to put the server into http boot mode). In a nutshell, http boot is really just PXE boot with http instead of tftp. This will allow to directly boot the full image from http(s), without having to chain-load it via TFTP => iPXE. Larger images are allowed, and the transmission can be secured with TLS (provided that the admin correctly deploys the certificates via Redfish as well).
2. Set certain parameters on the system that can later be read from the RAM disk image (e.g. we could use that to deploy an activation key grain to the firmware I guess, or set a hostname that the image can pick up).
3. At least in some vendor extensions, you can directly point Redfish to a URL with the boot media, no need for PXE or http boot.

Finally, you can also switch the system on via Redfish and do firmware updates or configure RAID controllers.

If we really want to go all in with my idea of having a fully idempotent "deployment template" from firmware/RAID to boot to Yomi to "normal" high state, yes, we'll need to do some work (mostly in the Salt stack itself).

Reply:
I integrated this description in the RFC.

> If we really want to go all in with my idea of having a fully idempotent "deployment template" from firmware/RAID to boot to Yomi to "normal" high state, yes, we'll need to do some work (mostly in the Salt stack itself).

I think that I get this idea, but do you have this written in any other place? I think that will help this RFC, and can impact a lot into Yomi

Reply:
I have prepared a few slides for a partner, trying to explain the concept. Will send them to you. But this idea certainly needs more work. Just to give you an idea: I have no clue whether that "master template" should actually run a highstate or whether it would be more useful to introduce an "ultra-highstate" (name TBD) that runs all three of (hardware prep, Yomi, normal highstate) as required, and you could still use them separately as well.